File: android/gemcore/src/main/kotlin/com/gemwallet/android/domains/name/AddressInputResolving.kt
package com.gemwallet.android.domains.name

import com.wallet.core.primitives.Chain
import com.wallet.core.primitives.NameRecord
import uniffi.gemstone.GemRecipient
import uniffi.gemstone.GemRecipientValidation

interface AddressInputResolving {
    fun validateRecipient(chain: Chain, input: String, nameRecord: NameRecord?): GemRecipientValidation

    fun recipient(chain: Chain, input: String, nameRecord: NameRecord?, memo: String?, references: List<String>): GemRecipient

    fun isNameSupported(name: String): Boolean

    suspend fun getNameRecord(name: String, chain: Chain): NameRecord?
}

File: ios/Packages/GemstonePrimitives/Sources/AddressInputResolving.swift
// Copyright (c). Gem Wallet. All rights reserved.

import class Gemstone.GemManageContactService
import class Gemstone.GemNameService
import class Gemstone.GemRecipientService
import struct Gemstone.GemRecipientValidation
import struct Gemstone.GemRecipient
import Primitives

public protocol AddressInputResolving: AnyObject, Sendable {
    func validateRecipient(chain: String, input: String, nameRecord: String?) -> GemRecipientValidation
    func recipient(chain: String, input: String, nameRecord: String?, memo: String?, references: [String]) throws -> GemRecipient
    func isNameSupported(name: String) -> Bool
    func getNameRecord(name: String, chain: String) async throws -> String?
}

public extension AddressInputResolving {
    func getNameRecord(name: String, chain: Primitives.Chain) async throws -> Primitives.NameRecord? {
        try await getNameRecord(name: name, chain: chain.rawValue).map { try Primitives.NameRecord($0) }
    }
}

extension GemNameService: AddressInputResolving {}
extension GemRecipientService: AddressInputResolving {}
extension GemManageContactService: AddressInputResolving {}
